app/controllers/UploadController.scala

```diff
@@ -17,9 +19,30 @@ class UploadController @Inject()(val controllerComponents: SecurityComponents,
                                  val graphqlConfiguration: GraphQLConfiguration,
                                  val keycloakConfiguration: KeycloakConfiguration,
                                  val frontEndInfoConfiguration: FrontEndInfoConfiguration)
-                                (implicit val ec: ExecutionContext) extends ValidatedActions with I18nSupport {
+                                (implicit val ec: ExecutionContext) extends TokenSecurity with I18nSupport {
+
+  def uploadPage(consignmentId: UUID): Action[AnyContent] = secureAction.async { implicit request: Request[AnyContent] =>
+    val consignmentStatusService = new ConsignmentStatusService(graphqlConfiguration)
+
+    for {
+      consignmentStatus <- consignmentStatusService.consignmentStatus(consignmentId, request.token.bearerAccessToken)
+    } yield {
+      val transferAgreementStatus: Option[String] = consignmentStatus.flatMap(_.transferAgreement)
+      val uploadStatus: Option[String] = consignmentStatus.flatMap(_.upload)
 
-  def uploadPage(consignmentId: UUID): Action[AnyContent] = uploadPermitted(consignmentId) { implicit request: Request[AnyContent] =>
-    Ok(views.html.upload(consignmentId, frontEndInfoConfiguration.frontEndInfo)).uncache()
+      transferAgreementStatus match {
+        case Some("Completed") =>
+          uploadStatus match {
+            case Some("InProgress") =>
+              Ok(views.html.uploadInProgress(consignmentId))
+            case Some("Completed") =>
+              Ok(views.html.uploadHasCompleted(consignmentId))
```

Comment:
I missed this before, but all of the responses should have the cache-control header. Otherwise the user might return to the page and see the cached version (e.g. the InProgress one) rather than the latest one.

Reply:
👍